### PR TITLE
DSDEGP-1629: Clio UUIDs are not lexicographically sortable

### PR DESCRIPTION
**Pull request by tbl3rd**
Use Firebase Push ID instead of UUIDv1.
Rename ClioUUIDGenerator -=> UpsertIdGenerator and so on.
Updated tests and fixed what broke.

Please look at UpsertIdGenerator.
The rest of the changes are almost compiler-driven.

https://github.com/broadinstitute/clio/blob/tbl-DSDEGP-1629-upsertids-do-not-sort/clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/util/UpsertIdGenerator.scala

### Description

_Please explain the changes you made here._
_Explain the **motivation** for making this change. What existing problem does the pull request solve?_
_Mention any JIRA issues fixed, addressed or otherwise related to this pull request, including issue numbers or hard links for issues in other repos._

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: Ensure that you have added or updated tests
- [ ] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509).
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Move the JIRA issue to QA once this checklist is completed

### CODE REVIEW

File: clio-dataaccess-model/src/test/scala/org/broadinstitute/clio/server/dataaccess/util/UpsertIdGeneratorSpec.scala
Comment (danxmoran) [inline clio-dataaccess-model/src/test/scala/org/broadinstitute/clio/server/dataaccess/util/UpsertIdGeneratorSpec.scala]: Should remove the println before merging.

File: clio-dataaccess-model/src/test/scala/org/broadinstitute/clio/server/dataaccess/util/UpsertIdGeneratorSpec.scala
Comment (danxmoran) [inline clio-dataaccess-model/src/test/scala/org/broadinstitute/clio/server/dataaccess/util/UpsertIdGeneratorSpec.scala]: Wondering: do you think it'd be a realistic test of the generator's thread-safety if we made this generate IDs inside `Future`s, then wrapped the whole thing inside a `Future.sequence`?

**Review by danxmoran — APPROVED** [review]

**Comment by tbl3rd** [comment]
I have a note to remove com.fasterxml.uuid if possible.

File: clio-dataaccess-model/src/test/scala/org/broadinstitute/clio/server/dataaccess/util/UpsertIdGeneratorSpec.scala
Comment (tbl3rd) [reply clio-dataaccess-model/src/test/scala/org/broadinstitute/clio/server/dataaccess/util/UpsertIdGeneratorSpec.scala]: Yep.

File: clio-dataaccess-model/src/test/scala/org/broadinstitute/clio/server/dataaccess/util/UpsertIdGeneratorSpec.scala
Comment (tbl3rd) [reply clio-dataaccess-model/src/test/scala/org/broadinstitute/clio/server/dataaccess/util/UpsertIdGeneratorSpec.scala]: I don't know.  I can try it.

File: clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ClioDocument.scala
Comment (pshapiro4broad) [inline clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ClioDocument.scala:15]: Maybe now would be a good time to define a type alias for the type of this field? Or, even better, use a custom type and get ourselves some type safety while we're at it.

File: clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/util/UpsertIdGenerator.scala
Comment (pshapiro4broad) [inline clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/util/UpsertIdGenerator.scala]: I think Scala conventions for static constants is to use upper camel case, e.g. `RandomCount`. Also it would be nice to have comments for the vals here.

**Review by pshapiro4broad — APPROVED** [review]
Can you also remove `ClioParsers.uuidParser` and `StateDecoder.decodeUUID` and the UUID case from `AutoElasticsearchIndex.getFieldDefinition()`?

File: clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/util/UpsertIdGenerator.scala
Comment (tbl3rd) [reply clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/util/UpsertIdGenerator.scala]: Their names are their comments.
Virtually everything is constant the way we write Scala here, but OK.
The object comment explains those 2 constants.

File: clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ClioDocument.scala
Comment (tbl3rd) [reply clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ClioDocument.scala:15]: Later. That's an internal fix we can do at any time. I want to unblock ClioP1.

And remember this?
https://broadinstitute.slack.com/archives/C2B5Q266P/p1506443608000468

(We should leave the UUID parsing in place as we're sure to need it soon.)

File: clio-dataaccess-model/src/test/scala/org/broadinstitute/clio/server/dataaccess/util/UpsertIdGeneratorSpec.scala
Comment (tbl3rd) [reply clio-dataaccess-model/src/test/scala/org/broadinstitute/clio/server/dataaccess/util/UpsertIdGeneratorSpec.scala]: Is this what you had in mind?
``` scala
  def generateIds(): Future[Seq[String]] = Future.sequence {
    Seq.fill(count) {
      Future(UpsertIdGenerator.nextId())
    }
  }
```
I'd have to move `TestKitSuite` somewhere accessible to the dataaccess module to write the tests though.

Comment (danxmoran) [reply clio-dataaccess-model/src/test/scala/org/broadinstitute/clio/server/dataaccess/util/UpsertIdGeneratorSpec.scala]: Yup that's what I was thinking!

I think all we'd need from the `TestKitSuite` for this is the execution context. Can you try using the default EC instead? `scala.concurrent.ExecutionContext.Implicits.global`

**Comment by tbl3rd** [comment]
Thanks for the review.

**Pull request merged**
